File: java-service/src/main/java/com/edu/platform/model/VideoLesson.java
package com.edu.platform.model;

import com.edu.platform.model.enums.VideoStatus;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "video_lessons", schema = "academic")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class VideoLesson {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // ==================== Relationships ====================
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "lesson_topic_id")
    private LessonTopic lessonTopic;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "teacher_id", nullable = false)
    private User teacher;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "subject_id", nullable = false)
    private Subject subject;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "session_recording_id")
    private SessionRecording sessionRecording;

    @OneToOne(mappedBy = "videoLesson", cascade = CascadeType.ALL, orphanRemoval = true)
    private VideoTranscript transcript;

    @OneToMany(mappedBy = "videoLesson", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<VideoChapter> chapters = new HashSet<>();

    @OneToMany(mappedBy = "videoLesson", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<VideoWatchHistory> watchHistories = new HashSet<>();

    // ==================== YouTube Details ====================
    
    @Column(name = "youtube_video_id", unique = true, length = 20)
    private String youtubeVideoId;

    @Column(name = "youtube_url", columnDefinition = "TEXT")
    private String youtubeUrl;

    @Column(name = "embed_url", columnDefinition = "TEXT")
    private String embedUrl;

    @Column(name = "youtube_channel_id", length = 50)
    private String youtubeChannelId;

    // ==================== Metadata ====================
    
    @Column(name = "title", nullable = false, length = 200)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "duration_seconds")
    private Integer durationSeconds;

    @Column(name = "thumbnail_url", columnDefinition = "TEXT")
    private String thumbnailUrl;

    @Column(name = "thumbnail_custom_url", columnDefinition = "TEXT")
    private String thumbnailCustomUrl;

    // ==================== Processing Status ====================
    
    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20)
    @Builder.Default
    private VideoStatus status = VideoStatus.PENDING;

    @Column(name = "upload_date")
    private Instant uploadDate;

    @Column(name = "processing_completed_at")
    private Instant processingCompletedAt;

    // ==================== AI-Generated Content Flags ====================
    
    @Column(name = "has_transcript")
    @Builder.Default
    private Boolean hasTranscript = false;

    @Column(name = "has_chapters")
    @Builder.Default
    private Boolean hasChapters = false;

    @Column(name = "has_quiz")
    @Builder.Default
    private Boolean hasQuiz = false;

    @Column(name = "has_summary")
    @Builder.Default
    private Boolean hasSummary = false;

    // ==================== Access Control ====================
    
    @Column(name = "is_aspirant_material")
    @Builder.Default
    private Boolean isAspirantMaterial = false;

    @Column(name = "is_public")
    @Builder.Default
    private Boolean isPublic = false;

    @Column(name = "is_premium")
    @Builder.Default
    private Boolean isPremium = false;

    // ✅ NEW FIELD: Published status (teacher can control visibility)
    @Column(name = "published")
    @Builder.Default
    private Boolean published = false;

    // ==================== Analytics Summary ====================
    
    @Column(name = "total_views")
    @Builder.Default
    private Integer totalViews = 0;

    @Column(name = "average_completion_rate", precision = 5, scale = 2)
    private BigDecimal averageCompletionRate;

    // ==================== Audit Fields ====================
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
        if (this.uploadDate == null) {
            this.uploadDate = Instant.now();
        }
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = Instant.now();
    }

    // ==================== Helper Methods ====================
    
    /**
     * Alias for teacher (used in VideoProcessingOrchestrator)
     */
    public User getUploadedBy() {
        return this.teacher;
    }

    public void setUploadedBy(User user) {
        this.teacher = user;
    }

    /**
     * Set upload date (used in VideoProcessingOrchestrator)
     */
    public void setUploadedAt(Instant uploadedAt) {
        this.uploadDate = uploadedAt;
    }

    public Instant getUploadedAt() {
        return this.uploadDate;
    }
}

File: python-service/app/models/video_lesson.py
from sqlalchemy import Column, Integer, String, Text, DateTime, Boolean, ForeignKey, Enum as SQLEnum, Numeric, func
from sqlalchemy.orm import relationship
from app.models.base import Base
import enum
from datetime import datetime


class VideoStatus(str, enum.Enum):
    PENDING = "PENDING"
    PROCESSING = "PROCESSING"
    PUBLISHED = "PUBLISHED"
    FAILED = "FAILED"


class VideoLesson(Base):
    """VideoLesson model - matches Java VideoLesson entity"""
    __tablename__ = "video_lessons"
    __table_args__ = {'schema': 'academic'}

    id = Column(Integer, primary_key=True, index=True)

    # Foreign Keys
    # ✅ KEEP ForeignKey for same-schema references (academic.*)
    lesson_topic_id = Column(Integer, ForeignKey('academic.lesson_topics.id', ondelete='SET NULL'))
    subject_id = Column(Integer, ForeignKey('academic.subjects.id', ondelete='CASCADE'), nullable=False)
    session_recording_id = Column(Integer, ForeignKey('academic.session_recordings.id', ondelete='SET NULL'))
    
    # ✅ NO ForeignKey for cross-schema reference (core.users) - just a plain column
    # The constraint exists at DB level but SQLAlchemy won't try to resolve it
    teacher_id = Column(Integer, nullable=False)

    # Relationships
    lesson_topic = relationship("LessonTopic", lazy="select", foreign_keys=[lesson_topic_id])
    subject = relationship("Subject", lazy="select", foreign_keys=[subject_id])
    session_recording = relationship("SessionRecording", back_populates="video_lesson", lazy="select", foreign_keys=[session_recording_id])
    transcript = relationship("VideoTranscript", back_populates="video_lesson", uselist=False, cascade="all, delete-orphan")
    chapters = relationship("VideoChapter", back_populates="video_lesson", cascade="all, delete-orphan")
    watch_histories = relationship("VideoWatchHistory", back_populates="video_lesson", cascade="all, delete-orphan")

    # YouTube Details
    youtube_video_id = Column(String(20), unique=True)
    youtube_url = Column(Text)
    embed_url = Column(Text)
    youtube_channel_id = Column(String(50))

    # Metadata
    title = Column(String(200), nullable=False)
    description = Column(Text)
    duration_seconds = Column(Integer)
    thumbnail_url = Column(Text)
    thumbnail_custom_url = Column(Text)

    # Processing Status
    status = Column(SQLEnum(VideoStatus), default=VideoStatus.PENDING)
    upload_date = Column(DateTime(timezone=True), server_default=func.now())
    processing_completed_at = Column(DateTime(timezone=True))

    # AI-Generated Content Flags
    has_transcript = Column(Boolean, default=False)
    has_chapters = Column(Boolean, default=False)
    has_quiz = Column(Boolean, default=False)
    has_summary = Column(Boolean, default=False)

    # Access Control
    is_aspirant_material = Column(Boolean, default=False)
    is_public = Column(Boolean, default=False)
    is_premium = Column(Boolean, default=False)
    published = Column(Boolean, default=False)

    # Analytics Summary
    total_views = Column(Integer, default=0)
    average_completion_rate = Column(Numeric(5, 2))

    # Audit Fields
    created_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=datetime.utcnow)

    def __repr__(self):
        return f"<VideoLesson(id={self.id}, title={self.title}, youtube_id={self.youtube_video_id})>"